### PR TITLE
Truncate user and group ids for 64 bit Linux systems

### PR DESCRIPTION
**Pull request by klingtnet**
Calls to `getuid`, `getgid` and their `eid` variants fail to compile on
64bit Linux systems because the return value of the syscall is of
`usize` and needs to be truncated to fit the size of `uid_t` that is 32
bit.

Thanks to @FireFox317 for figuring this out in Zig's Discord channel!

**TODO**

- [ ] squash commits
- [x] fix CI test
- [x] run regression test

### CODE REVIEW

File: lib/std/os/linux.zig [hunk 1]
```diff
@@ -747,31 +747,31 @@ pub fn getuid() uid_t {
     if (@hasField(SYS, "getuid32")) {
         return @as(uid_t, syscall0(.getuid32));
     } else {
-        return @as(uid_t, syscall0(.getuid));
+        return @as(uid_t, @truncate(u32, syscall0(.getuid)));
```
Comment (FireFox317) [inline lib/std/os/linux.zig:750]:
```suggestion
        return @truncate(uid_t, syscall0(.getuid));
```

File: lib/std/os/linux.zig [hunk 2]
```diff
@@ -747,31 +747,31 @@ pub fn getuid() uid_t {
     if (@hasField(SYS, "getuid32")) {
         return @as(uid_t, syscall0(.getuid32));
```
Comment (FireFox317) [inline lib/std/os/linux.zig:748]:
```suggestion
        return @truncate(uid_t, syscall0(.getuid32));
```
Could you fix these too?

**Review by FireFox317 — CHANGES_REQUESTED** [review]
Thanks for the fast PR @klingtnet ! I have a couple suggestions.

File: lib/std/os/linux.zig (hunk 1 above)
Comment (klingtnet) [reply lib/std/os/linux.zig]: Thanks, wasn't sure where to place the truncate.  I adjusted all changes accordingly.

File: lib/std/os/linux.zig (hunk 2 above)
Comment (klingtnet) [reply lib/std/os/linux.zig:748]: As mentioned above, addressed all occurrences.

File: lib/std/os/linux.zig (hunk 1 above)
Comment (klingtnet) [reply lib/std/os/linux.zig]: Need, to remove the redundant `@as`, forgot about that.

Comment (klingtnet) [reply lib/std/os/linux.zig]: Looks fine from my side now.

**Comment by LemonBoy** [comment]
What about a nice test in `os/linux/test.zig` so that it won't accidentally regress?
Something like:

```
test "" {
    expectEqual(getauxval(AT_UID), getuid());
    // same for AT_GID, AT_EUID, AT_EGID
}
```

File: lib/std/os/linux.zig [hunk 3]
```diff
@@ -745,33 +745,33 @@ pub fn setregid(rgid: gid_t, egid: gid_t) usize {
 
 pub fn getuid() uid_t {
     if (@hasField(SYS, "getuid32")) {
-        return @as(uid_t, syscall0(.getuid32));
+        return @truncate(uid_t, syscall0(.getuid32));
```
Comment (daurnimator) [inline lib/std/os/linux.zig:748]: Shouldn't this be an `@intCast`? We don't actually expect values larger than `uid_t`.

**Review by daurnimator — CHANGES_REQUESTED** [review]

**Comment by klingtnet** [comment]
> What about a nice test in `os/linux/test.zig` so that it won't accidentally regress?
> Something like:
> 
> ```
> test "" {
>     expectEqual(getauxval(AT_UID), getuid());
>     // same for AT_GID, AT_EUID, AT_EGID
> }
> ```

Sounds fine to me, will address this after work.

**Comment by klingtnet** [comment]
I added a regression test as suggested, unfortunately I could not find any documentation on how to run them. 

If it does not exist already then there should be some documentation for new contributors.

**Comment by LemonBoy** [comment]
> I added a regression test as suggested, unfortunately I could not find any documentation on how to run them.
> 
> If it does not exist already then there should be some documentation for new contributors.

You can simply run `zig test <your .zig file>`, that's much faster than running the whole test suite (that you can run with `zig build --build-file ../build.zig test-std`, YMMV).

**Comment by klingtnet** [comment]
> > I added a regression test as suggested, unfortunately I could not find any documentation on how to run them.
> > If it does not exist already then there should be some documentation for new contributors.
> 
> You can simply run `zig test <your .zig file>`, that's much faster than running the whole test suite (that you can run with `zig build --build-file ../build.zig test-std`, YMMV).

Thank you very much!  Test's will probably fail with `0.7.1` so I need to wait until my fresh build is ready (required to install `llvm` and `lld`, luckily the latter one was mentioned in https://github.com/ziglang/zig/issues/419 ).

**Comment by LemonBoy** [comment]
> The test does not compile with 0,7.1:

Yes, some tests are not meant to be run stand-alone. You want to run `lib/std/os.zig` here.

File: lib/std/os/linux.zig (hunk 3 above)
Comment (klingtnet) [reply lib/std/os/linux.zig]: Sounds reasonable, replaced `@truncate` with `@intCast`.

**Comment by klingtnet** [comment]
So, to run the tests locally I needed to compile `zig` locally under Arch Linux.  I will list the important steps here for documentation purposes:

```sh
$ pacman -S llvm clang lld
# follow the wiki instructions
$ cmake .. -DZIG_PREFER_CLANG_CPP_DYLIB=True
# wait ... done
```

The test still does not compile:

```sh
$ build/zig test lib/std/os/linux/test.zig
./lib/std/os/linux/test.zig:105:17: error: use of undeclared identifier 'getauxval'
    expectEqual(getauxval(AT_UID), getuid());
                ^
```

I imported `getauxval` from `std.os.linux` and then it fails with:

```sh
$ build/zig test lib/std/os/linux/test.zig | clipit
./lib/std/os/linux/test.zig:105:17: error: use of undeclared identifier 'getauxval'
    expectEqual(getauxval(AT_UID), getuid());
                ^
```

Unfortunately, I could not find a reference to `AT_UID` to import it.

**Comment by LemonBoy** [comment]
> I imported getauxval from std.os.linux and then it fails with:

You can use `linux.getauxval` instead of defining an extra const for something that's only used once.

> I imported getauxval from std.os.linux and then it fails with:

Those are defined in `std.elf`... It makes no sense, I know.

**Comment by klingtnet** [comment]
Seems to work fine now:

```sh
$ build/zig test lib/std/os/linux/test.zig
All 5 tests passed.
s1 alinz@kn-portable
```

**Comment by klingtnet** [comment]
I need to fix the [failing CI test](https://dev.azure.com/ziglang/0184afc2-b8d3-409b-8e4f-4032aaca4c96/_build/results?buildId=11195):

```
1081/1637 os.linux.test.test "std-native-Debug-c-multi user and group ids"... expected 0, found 1001
/home/vsts/work/1/s/lib/std/testing.zig:74:32: 0x44178f in testing.expectEqual (test)
                std.debug.panic("expected {}, found {}", .{ expected, actual });
                               ^
/home/vsts/work/1/s/lib/std/os/linux/test.zig:105:16: 0x38807e in os.linux.test.test "std-native-Debug-c-multi user and group ids" (test)
    expectEqual(linux.getauxval(elf.AT_UID), linux.getuid());
               ^
/home/vsts/work/1/s/lib/std/special/test_runner.zig:61:28: 0x3c0a58 in std.special.main (test)
        } else test_fn.func();
                           ^
/home/vsts/work/1/s/lib/std/start.zig:344:37: 0x3c27a1 in start.main (test)
            const result = root.main() catch |err| {
                                    ^
error: the following test command crashed:
/home/vsts/work/1/s/zig-cache/o/590cc85149a19cdf50140d5408f6d1bb/test
The following command exited with error code 1:
/home/vsts/work/1/s/build/zig test /home/vsts/work/1/s/lib/std/std.zig -lc --test-name-prefix std-native-Debug-c-multi  --cache-dir /home/vsts/work/1/s/zig-cache --global-cache-dir /home/vsts/.cache/zig --name test -I /home/vsts/work/1/s/test --override-lib-dir /home/vsts/work/1/s/lib 
error: the following build command failed with exit code 1:
/home/vsts/work/1/s/zig-cache/o/b44c2f8e472997c666788c7413e997b6/build /home/vsts/work/1/s/build/zig /home/vsts/work/1/s /home/vsts/work/1/s/zig-cache /home/vsts/.cache/zig test -Denable-qemu -Denable-wasmtime
```

**Comment by FireFox317** [comment]
The problem is in the `getauxval` function:
https://github.com/ziglang/zig/blob/506af7e52e0985b410ea089bf5fa3247ab2377cb/lib/std/os/linux.zig#L37-L49

As you can see the `elf_aux_maybe` is populated by the `start.zig` file and I'm pretty sure this startup code is not run when we do `zig test`. So it looks like we cannot test this (or somebody else knows a solution to this).

**Comment by LemonBoy** [comment]
Oh this is some other `std.os` bullshit, slap a `if (builtin.link_libc) return error.SkipZigTest;` and let's call it a day, this is a test for our direct syscall wrappers after all.

**Pull request merged**